test/utils.py

```diff
@@ -0,0 +1,373 @@
+        self.key = 0
+        self.tasks: dict[int, _Task] = {}
+
+    def _create_processes(self):
+        return [_ProcessWrapper(self, i)
+                for i in range(self.n_processes)]
+
+    def _kill_and_restart(self, i: int):
+        self.processes[i].p.kill()
+        self.processes[i] = _ProcessWrapper(self, i)
+
+    async def _wait_for_empty_process(self, interval: float = 0):
+        while True:
+            for i, p in enumerate(self.processes):
+                self._update_process(i)
+                if p.is_waiting():
+                    return i, p
+            await asyncio.sleep(interval)
+
+    def _update_process(self, i: int):
+        p = self.processes[i]
+        if p.is_dead():
+            self._kill_and_restart(i)
+        p.update_output_status()
+
+    def _update_all_processes(self):
+        for i, _ in enumerate(self.processes):
+            self._update_process(i)
+
+    async def apply(self, fn, args, kwargs, timeout: float = None, interval: float = 0):
+        key = self.key
+        self.key += 1
+        task = self.tasks[key] = _Task((key, fn, args, kwargs))
+        # Only submit one task at a time, avoids being killed
+        # because something else timed out, etc.
+        idx, p = await self._wait_for_empty_process(interval)
+        p.submit(task)
+        start = time.perf_counter()
+        while self.tasks[key].output is None:
+            if timeout and time.perf_counter() > start + timeout:
+                self._kill_and_restart(idx)
+                del p  # I don't feel the need to keep that in tb_frame.f_locals forever
+                raise TimeoutError
+            await asyncio.sleep(interval)
+            self._update_all_processes()
+        output = self.tasks[key].output
+        del self.tasks[key]  # Don't leak memory
+        _, success, value_or_err = output
+        if success:
+            return value_or_err
+        else:
+            raise value_or_err
+
+
+class _ProcessWrapper:
+    def __init__(self, parent: SimpleProcessPool, i: int):
+        self.i = i
+        self.pool = parent
+        self.tasks_in = mp.Queue()
+        self.results_out = mp.Queue()
+        self.p = mp.Process(name=f'_SimpleProcessPool:worker-{i}',
+                            target=self._worker,
+                            args=(self.tasks_in, self.results_out),
+                            daemon=True)
+        self.waiting = True
+        self.p.start()
+
+    def submit(self, task: _Task):
+        self.waiting = False
+        self.tasks_in.put(task.inputs)
+
+    def is_dead(self):
+        return not self.p.is_alive()
+
+    def is_waiting(self):
+        self.update_output_status()
+        return self.p.is_alive() and self.waiting
+
+    def update_output_status(self):
+        try:
+            result = self.results_out.get(False)
+        except queue.Empty:
+            return
+        self.waiting = True
+        key, *_ = result
+        self.pool.tasks[key].output = result
```

Comment:
Messing with `.pool`'s attributes isn't the nicest way of doing this... a method in `SimpleProcessPool` would be better.